Sources/Apollo/WebSocketTransport.swift

```diff
@@ -0,0 +1,342 @@
+        
+        if let error = self.error {
+            completionHandler(nil,error)
+        }
+        
+        return WebSocketTask(self,operation) { (body, error) in
+            if let body = body {
+                let response = GraphQLResponse(operation: operation, body: body)
+                completionHandler(response,error)
+            } else {
+                completionHandler(nil,error)
+            }
+        }
+        
+    }
+    
+    public func isConnected() -> Bool {
+        return websocket?.isConnected ?? false
+    }
+        
+    private func processMessage(socket: WebSocketClient, text: String) {
+        
+        if let data = text.data(using: (.utf8) ) {
+            do {
+                let json = try JSONSerializationFormat.deserialize(data: data) as? JSONObject
+                if let msgtype = json?["type"] as? String {
+                    switch(msgtype) {
+                    case OperationMessage.Types.DATA.rawValue:
+                        if let id = json?["id"] as? String,
+                            let responseHandler = subscribers[id] {
+                            
+                            let payload = json?["payload"] as? JSONObject
+                            responseHandler(payload,nil)
+                        }
+                        
+                    case OperationMessage.Types.ERROR.rawValue:
+                        if let id = json?["id"] as? String,
+                            let responseHandler = subscribers[id] {
+                            
+                            let payload = json?["payload"] as? JSONObject
+                            let responseError = json?["error"] as? Error
+                            let error = WebSocketError(payload: payload, error: responseError, kind: .errorResponse)
+                            responseHandler(payload,error)
+                        }
+                        
+                    case OperationMessage.Types.COMPLETE.rawValue:
+                        if let id = json?["id"] as? String {
+                            // remove the callback if NOT a subscription
+                            if subscriptions[id] == nil {
+                                subscribers.removeValue(forKey: id)
+                            }
+                        }
+                        
+                    case OperationMessage.Types.CONNECTION_ACK.rawValue,
+                         OperationMessage.Types.CONNECTION_KEEP_ALIVE.rawValue:
+
+                        acked = (msgtype == OperationMessage.Types.CONNECTION_ACK.rawValue)
+                        writeQueue()
+
+                    default:
+                        print("WebSocketTransport::unprocessed event \(msgtype)")
+                    }
+                }
+            } catch {
+                print("WebSocketTransport::unprocessed event \(data)")
+            }
+        } else {
+            print("WebSocketTransport::unprocessed event \(text)")
+        }
```

Comment:
This function is quite big. It seems to do two things: Parse a message object from a string, and act depending on what message it is.

I think it would make sense to have the parsing logic in `OperationMessage` and make `id`, `error` and `payload` associated values to the different types to avoid `if let` logic here (what happens if you get a `COMPLETE` message but `id` is `nil`?).

It would then make the flow much easier to test and understand.

Reply:
Something like this? 

   private func processMessage(socket: WebSocketClient, text: String) {
        
        OperationMessage.parseMessage(text) { (type,id,payload,error) in
            if let type = type {
                switch(type) {
                    case OperationMessage.Types.DATA.rawValue,
                         OperationMessage.Types.ERROR.rawValue:
                        if let id = id {
                            if let responseHandler = subscribers[id] {
                                responseHandler(payload,error)
                            }
                        } else {
                            print("WebSocketTransport::unprocessed message \(text)")
                        }
                    
                    case OperationMessage.Types.COMPLETE.rawValue:
                        if let id = id {
                            // remove the callback if NOT a subscription
                            if subscriptions[id] == nil {
                                subscribers.removeValue(forKey: id)
                            }
                        } else {
                            print("WebSocketTransport::unprocessed message \(text)")
                        }
                        
                    case OperationMessage.Types.CONNECTION_ACK.rawValue,
                         OperationMessage.Types.CONNECTION_KEEP_ALIVE.rawValue:
                        
                        acked = (type == OperationMessage.Types.CONNECTION_ACK.rawValue)
                        writeQueue()
                        
                    default:
                        print("WebSocketTransport::unprocessed message \(text)")
                    }
            } else {
                print("WebSocketTransport::unprocessed message \(text)")
            }
        }
    }